projects/output-and-error-handling/main.go

```diff
@@ -0,0 +1,111 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"strconv"
+	"time"
+)
+
+func handleWeatherRequest() (string, error) {
+
+	resp, err := http.Get("http://localhost:8080")
+
+	if err != nil {
+
+		if os.IsTimeout(err) {
+			fmt.Fprintf(os.Stderr ,"request timed out.\n");
+			return "", fmt.Errorf("request timed out")
+		}
+		if err == io.EOF {
+			fmt.Fprintf(os.Stderr ,"connection terminated unexpectedly.\n")
+			return "", fmt.Errorf("connection terminated unexpectedly")
+		}
+		fmt.Fprintf(os.Stderr ,"failed to fetch server: %v\n", err);
+		return "", fmt.Errorf("failed to fetch server: %v", err)
+	}
+
+	defer resp.Body.Close()
+
+	if resp.StatusCode == http.StatusTooManyRequests {
+		retryAfter := resp.Header.Get("Retry-After")
+
+		if retryAfter == "" {
+			fmt.Fprintf(os.Stderr,"server is too busy")
+
+			// 2 seconds is a good choice because it's short enough to minimize user delay while still giving the server a
+			//  brief moment to recover, avoiding overloading it with frequent requests.
+			time.Sleep(2 * time.Second)
+			return handleWeatherRequest()
+		} else {
+			retryAfterInSecond, err := strconv.Atoi(retryAfter)
+			if err != nil {
+
+				retryTime,err:= time.Parse(time.RFC1123, retryAfter);
+
+				if err!=nil{
+					fmt.Fprintf(os.Stderr, "error happened: %v\n" , err);
+					time.Sleep(2 * time.Second)
+					return handleWeatherRequest()
+				}
+
+				waitingDuration:=time.Until(retryTime);
+				if waitingDuration.Seconds() > 5 {
+
+					fmt.Fprintf(os.Stderr,"server response takes %.0f seconds, process canceled", waitingDuration.Seconds())
+
+					return "", fmt.Errorf("server response takes %.0f seconds, process canceled", waitingDuration.Seconds())
+
+				}else if waitingDuration.Seconds()>0{
+
+					fmt.Fprintf(os.Stderr ,"Waiting for %.0f seconds...\n", waitingDuration.Seconds())
+
+					time.Sleep(waitingDuration);
+					return handleWeatherRequest();
+				}else{
+					fmt.Fprintf(os.Stderr,"retrying to get the weather... please wait")
+				}
+
+				return handleWeatherRequest();
+			}
```

Comment:
This whole "Working out how long to sleep before retrying" thing feels like it would be a great function to extract. It's a pure function (the same inputs always map to the same outputs), it has a really clear name for its purpose (maybe `parseRetryAfterHeader` or something), and would be really easy to test on its own. The current function reading:

```go
// ... stuff from before
timeToSleep, err := parseRetryAfterHeader(resp.Header.Get("Retry-After"))
if err != nil {
  // Log and return an error or something, maybe
}
time.Sleep(timeToSleep)
```

feels a lot easier to to read and understand what's going on than reading this whole block of code :)